server/datastore/mysql/operating_systems.go

```diff
@@ -0,0 +1,144 @@
+	case errors.Is(err, sql.ErrNoRows):
+		return newOperatingSystemDB(ctx, tx, hostOS)
+	default:
+		return nil, ctxerr.Wrap(ctx, err, "get operating system")
+	}
+}
+
+// `newOperatingSystemDB` inserts a record for the given operating system and
+// returns the record including the newly associated ID.
+func newOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	stmt := "INSERT INTO operating_systems (name, version, arch, kernel_version) VALUES (?, ?, ?, ?)"
+	if _, err := tx.ExecContext(ctx, stmt, hostOS.Name, hostOS.Version, hostOS.Arch, hostOS.KernelVersion); err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "insert new operating system")
+	}
+
+	switch storedOS, err := getOperatingSystemDB(ctx, tx, hostOS); {
+	case err == nil:
+		return storedOS, nil
+	case errors.Is(err, sql.ErrNoRows):
+		return nil, doRetryErr
+	default:
+		return nil, ctxerr.Wrap(ctx, err, "get new operating system")
+	}
+}
+
+// getOperatingSystemDB queries the `operating_systems` table with the
+// name, version, arch, and kernel_version of the given operating system.
+// If found, it returns the record including the associated ID.
+func getOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	var os fleet.OperatingSystem
+	stmt := "SELECT * FROM operating_systems WHERE name = ? AND version = ? AND arch = ? AND kernel_version = ? LIMIT 1"
+	if err := sqlx.GetContext(ctx, tx, &os, stmt, hostOS.Name, hostOS.Version, hostOS.Arch, hostOS.KernelVersion); err != nil {
+		return nil, err
+	}
+	return &os, nil
+}
+
+// maybeUpdateHostOperatingSystemDB checks the `host_operating_system` table
+// for the given host ID. If found, it compares the given operating system ID
+// against the stored operating system ID and updates the table accordingly.
+// If not found, it inserts a new record for the given host id with the given
+// operating system id.
+func maybeUpdateHostOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostID uint, osID uint) error {
+	switch storedID, err := getIDHostOperatingSystemDB(ctx, tx, hostID); {
+	case storedID == osID:
+		// ok
+	case errors.Is(err, sql.ErrNoRows):
+		if err := insertHostOperatingSystemDB(ctx, tx, hostID, osID); err != nil {
+			return ctxerr.Wrap(ctx, err, "insert host operating system")
+		}
+	case storedID != osID:
+		if err := updateHostOperatingSystemDB(ctx, tx, hostID, osID); err != nil {
+			return ctxerr.Wrap(ctx, err, "update host operating system")
+		}
+	default:
+		return ctxerr.Wrap(ctx, err, "get host operating system")
+	}
+	return nil
+}
+
+// getIDHostOperatingSystemDB queries the `host_operating_system` table and returns the
+// operating system ID for the given host ID.
+func getIDHostOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostID uint) (uint, error) {
+	var id uint
+	stmt := "SELECT os_id FROM host_operating_system WHERE host_id = ? LIMIT 1"
```

Comment:
DO we need `LIMIT 1`? It looks like a host can only ever be associated with one operating system.

Reply:
Ah, I was thinking that there would be a performance benefit with the `LIMIT 1` but on further research it looks like it wouldn't add any benefit when the `WHERE` clause is operating on the primary key. Is that consistent with your understanding?

Reply:
yes